src/main/core/logger/shadow_logger.rs

```diff
@@ -0,0 +1,426 @@
+        // Integers secs, after integer minutes are subtracted out.
+        let secs_part = whole_secs - u64::from(whole_mins) * 60;
+        // Nanos, after integer secs are subtracted out.
+        let nanos_part =
+            u64::try_from(total_nanos - u128::from(whole_secs) * 1_000_000_000).unwrap();
+
+        Self {
+            hours: whole_hours,
+            mins: mins_part,
+            secs: secs_part,
+            nanos: nanos_part,
+        }
+    }
+}
+
+#[cfg(test)]
+#[test]
+fn test_time_parts() {
+    assert_eq!(
+        TimeParts::from_nanos(
+            (Duration::from_nanos(1) + Duration::from_secs(3600 + 60 + 1)).as_nanos()
+        ),
+        TimeParts {
+            hours: 1,
+            mins: 1,
+            secs: 1,
+            nanos: 1
+        }
+    );
+}
+
+/// Initialize the Shadow logger.
+pub fn init() -> Result<(), SetLoggerError> {
+    log::set_logger(&*SHADOW_LOGGER)?;
+
+    // Start the thread that will receive log records and flush them to output.
+    std::thread::Builder::new()
+        .name("shadow-logger".to_string())
+        .spawn(move || SHADOW_LOGGER.logger_thread_fn())
+        .unwrap();
+
+    // Arrange to flush the logger on panic.
+    let default_panic_handler = std::panic::take_hook();
+    std::panic::set_hook(Box::new(move |panic_info| {
+        // Attempt to flush the logger. We want to avoid a recursive panic, so
+        // we flush the queue on the current thread instead of trying to send
+        // a command to the logger thread (because our thread-local sender
+        // may have already been destructed, and because the logger thread
+        // itself may be in a bad state), and ignore errors.
+        SHADOW_LOGGER.flush_records(None).ok();
+        default_panic_handler(panic_info);
+    }));
+
+    Ok(())
+}
+
+/// A logger specialized for Shadow. It attaches simulation context to log
+/// entries (e.g. sim time, running process, etc.). It's also designed for
+/// high performance to accomodate heavy logging from multiple threads.
+pub struct ShadowLogger {
+    // Channel used to send commands to the logger's thread.
+    //
+    // The Sender half of a channel isn't Sync, so we must protect it with a
+    // Mutex to make ShadowLogger be Sync. This is only accessed once per
+    // thread, though, to clone into the thread-local SENDER.
+    command_sender: Mutex<Sender<LoggerCommand>>,
+
+    // Like the sender, needs a Mutex for ShadowLogger to be Sync.
+    // The Mutex is only locked once though by the logger thread, which keeps
+    // it locked for as long as it's running.
+    command_receiver: Mutex<Receiver<LoggerCommand>>,
+
+    // A lock-free queue for individual log records. We don't put the records
+    // themselves in the `command_sender`, because `Sender` doesn't support
+    // getting the queue length. Conversely we don't put commands in this queue
+    // because it doesn't support blocking operations.
+    records: SegQueue<ShadowLogRecord>,
```

Comment:
Since the logger doesn't push more than `SYNC_FLUSH_QD_LINES_THRESHOLD` records, does it make sense to make this an `ArrayQueue` to avoid the extra allocation? Maybe this doesn't matter since there are already a few other string allocations per record.

Reply:
Yeah, I think I originally went with SegQueue because I started with a higher value for ASYNC_FLUSH_QD_LINES_THRESHOLD, and didn't want the fixed memory cost. With the current value of 100k it's not so bad, though.